### PR TITLE
Prebuild issues roundup

### PR DESCRIPTION
**Pull request by geropl**
## Description

Hey @filiptronicek :wave: 

This PR was meant to address a lot of the current prebuild issues, but I did not mange to finish it last week. :cry: 
The key piece missing is that we render the log output 2 or 3 times now, so a react invalidation problem:
 - I think the strategy of caching it by "prebuildId + taskId" in the respective tab's state can still work. But I was running out of time debugging where the invalidation&re-renders came from.
 - If that turns out to be too hard (this was where some of the referenced bug came from: it would not invalidate e.g. on prebuildId change :point_down: ), we can easily turn to caching in in localstorage - or even re-stream on every render and rely on HTTP caching - worked astonishingly good on the old `PrebuildLogs.tsx`)

Anyway, happy to sync for ~30mins a slot tomorrow morning if that helps. Also, happy to do the review! :partying_face: 

## Related Issue(s)

Meant to:
 - fixes ENT-282 (fixes streaming bug in server by avoiding duplicating frontend logic in backend)
 - fixes ENT-323 (shows SHA + duration)
 - fixes ENT-368 (re-purpose un-unsed QUEUE staate for everything before RUNNING)
 - fixes ENT-428 (we did miss to invalidate some components/the tabs when the `prebuildId` changed - now the keys are "prebuildId + taskId")

## How to test

https://gpl-prebuild-cleanup.preview.gitpod-dev.com/workspaces

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - gpl-prebuild-cleanup</li>
	<li><b>🔗 URL</b> - <a href="https://gpl-prebuild-cleanup.preview.gitpod-dev.com/workspaces" target="_blank">gpl-prebuild-cleanup.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - gpl-prebuild-cleanup-gha.27254</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-gpl-prebuild-cleanup%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-dev-preview" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/workspace/headless-log-service.ts [hunk 1]
```diff
@@ -145,14 +145,20 @@ export class HeadlessLogService {
...
 
-    protected async supervisorListTasks(logCtx: LogContext, logEndpoint: HeadlessLogEndpoint): Promise<TaskStatus[]> {
+    protected async supervisorListTasks(
+        logCtx: LogContext,
+        logEndpoint: HeadlessLogEndpoint,
+        cancel?: (retry: boolean) => void,
+    ): Promise<TaskStatus[]> {
         if (logEndpoint.url === "") {
             // if ideUrl is not yet set we're too early and we deem the workspace not ready yet: retry later!
+            cancel?.(false);
```
Comment (filiptronicek) [inline components/server/src/workspace/headless-log-service.ts:161]: @geropl the comment above indicates we should retry in this case? Or are we handling that separately?

**Comment by mustard-mh** [comment]
Will review and test this PR later 🌮

**Review by mustard-mh — APPROVED** [review]
It's all working well in general, and we can do the rest in follow up.

1️⃣ If a prebuild has image-build, and image-build failed, prebuild detail page will stuck in `Pending` phase and no updates until user refreshes the page -> Turn into `Error`, message `Error: headless task failed: exit status 1` and labels `Duration` and `Stopped` (time) shown

Example context https://github.com/mustard-mh/test/tree/prebuild/image-build-failed

|Stuck in `Pending`| Refresh Page|
|:-:|:-:|
|SCR-20240709-upgu|SCR-20240709-upkx|

2️⃣ nit: Click button to re-run prebuild, it will show `Image build` task tab then disappear (should be a follow up?)

3️⃣ ~~Sometimes stopped (watch until it stopped) prebuild detail page doesn't show `Stopped` and `Duration` labels on top right until page refreshed~~ I could not reproduce it anymore

4️⃣ nit: as we will show commit SHA on prebuild detail page, we could also show such label on prebuilds list page?

5️⃣ Before image build workspace created and recorded in db, dashboard will show `inaccessible` content for a long time, while as an user, I don't know if it's really running or pending or never able to access that logs
image

6️⃣ Re-run prebuild will use the latest commit instead of that commit SHA shows on prebuild detail page (SHA changed)

**Comment by filiptronicek** [comment]
@mustard-mh many thanks for the review!

> 1️⃣ If a prebuild has image-build, and image-build failed, prebuild detail page will stuck in Pending phase and no updates until user refreshes the page -> Turn into Error, message Error: headless task failed: exit status 1 and labels Duration and Stopped (time) shown

I created ENT-445 for this, nice find!

> 2️⃣ nit: Click button to re-run prebuild, it will show Image build task tab then disappear (should be a follow up?)

This boils down to us being unable to know whether an image build is going to happen from the shape we get from the API. Not sure if there's any quick wins to be had here.

> 3️⃣ Sometimes stopped (watch until it stopped) prebuild detail page doesn't show Stopped and Duration labels on top right until page refreshed I could not reproduce it anymore

Interestingly enough, I ran into this as well, but when I tried to show it to @geropl this morning, it didn't happen 😆. I think it means there is for sure an issue, but I could not pin it down yet.

> 4️⃣ nit: as we will show commit SHA on prebuild detail page, we could also show such label on prebuilds list page?

It used to be like this, but I removed it from the list because it took a lot of space in a table where we don't have much of it available.

> 5️⃣ Before image build workspace created and recorded in db, dashboard will show inaccessible content for a long time, while as an user, I don't know if it's really running or pending or never able to access that logs

This should most likely be dealt with separately, as this PR does not change this behavior. I agree we should make this part more reliable, but it might not be worth the effort.

> 6️⃣ Re-run prebuild will use the latest commit instead of that commit SHA shows on prebuild detail page (SHA changed)

I think this is fine as:
1. It was always like this
2. The ref which will be used is written on the button

**Review by filiptronicek — APPROVED** [review]

**Pull request merged**